Add mingwW64-llvm cross-system.

shlevy opened the pull request:
This adds a new cross system targeting mingwW64 + ucrt using an llvm-based toolchain.

Similar in spirit to and inspired by @mstorsjo's https://github.com/mstorsjo/llvm-mingw.

Among other benefits of LLVM + ucrt, this will be necessary to build a GHC compiler targeting Windows in 9.4 or later.

shlevy committed:
mingw-w64: 9.0.0 -> 10.0.0

shlevy committed:
cross: Allow Windows toolchains to use ucrt as libc.

shlevy committed:
cc-wrapper: clang doesn't support -fPIC on Windows

shlevy committed:
cross: Add mingwW64-llvm cross-system.

shlevy committed:
llvm-bintools: Include aliases for windres and dlltool on Windows

shlevy committed:
bintools-wrapper: Don't pass -z to lld targeting Windows

shlevy committed:
libunwind: Fix build against compiler-rt-using clang

shlevy committed:
libcxxabi: Fix build against compiler-rt-using clang

shlevy committed:
libcxx/libcxxabi: Fix build on Windows with lld

shlevy commented:
Easiest reviewed commit-wise.

shlevy committed:
llvm-bintools: passthru targetPrefix.

This is expected e.g. by the GHC cross-compilation code

Ericson2314 commented on pkgs/os-specific/windows/mingw-w64/default.nix:
```diff
@@ -1,22 +1,34 @@
+  version = "10.0.0";
+
+  archFlags =
+    if stdenv.targetPlatform.isAarch32
+    then [ "--disable-lib32" "--disable-lib64" "--enable-libarm32" ]
+    else if stdenv.targetPlatform.isAarch64
+    then [ "--disable-lib32" "--disable-lib64" "--enable-libarm64" ]
+    else if stdenv.targetPlatform.isx86_32
+    then [ "--enable-lib32" "--disable-lib64" ]
+    else if stdenv.targetPlatform.isx86_64
+    then [ "--disable-lib32" "--enable-lib64" ]
+    else [];
```
I would use `lib.enableFeature` for these to avoid O(n^2) string literals.

shlevy replied:
Done.

Ericson2314 commented on pkgs/os-specific/windows/mingw-w64/default.nix:
```diff
@@ -14,7 +14,8 @@ let
 
-in stdenv.mkDerivation {
+  crt = stdenv.hostPlatform.libc;
+in assert crt == "ucrt" || crt == "msvcrt"; stdenv.mkDerivation {
```
I would instead do
```nix
meta.broken = !(lib.elem crt [ "msvcrt" "ucrt" ]);
```
for each of the `asset`s.

shlevy replied:
Done.

Ericson2314 commented:
This generally looks good to me. It looks like you made some attempt to backport changes to older LLVMs to keep things consistent? That's good. I asked @sternenseemann to review too. There's a lot of new passthru items and I think they are OK, but it's a good thing to get a second set of eyes upon.

shlevy commented:
@Ericson2314 llvm 14 is not yet working fully, so this actually uses 13 by default. I forward-ported the changes to 14 so they won't get lost as llvm gets upgraded

sternenseemann commented on pkgs/development/compilers/llvm/13/bintools/default.nix:
```diff
@@ -1,11 +1,17 @@
+in runCommand "llvm-binutils-${version}" {
+  preferLocalBuild = true;
+  passthru = {
+    isLLD = true;
```
I'm a bit torn on this one. I think it's alright and probably useful, *but* it should be introduced in a consistent manner.

My proposal would be to add this separately bundled with a similar disambiguator added for `cctools` and GNU `binutils` as well. `bintools-wrapper` should also add the missing disambiguators via elaboration.

For now `stdenv.hostPlatform.linker == "lld"` works just as well, with the slight disadvantage of inaccuracy if a non-default `bintools` is passed *manually*.

shlevy replied:
> My proposal would be to add this separately bundled with a similar disambiguator added for cctools and GNU binutils as well. bintools-wrapper should also add the missing disambiguators via elaboration.

I don't object to this as an end, but do we need to do it all at once?

> For now stdenv.hostPlatform.linker == "lld" works just as well

No it doesn't, as I've been testing stdenvs using lld which is not the default for the platform.

sternenseemann replied:
Doing all things at once was kind of my point. E.g. `clangUseLLVM` doesn't really work well outside of a platform that actually has `useLLVM = true` and  `linker = "lld"` which is why it's used nowhere in nixpkgs except in the cross stdenv. In fact all LLVM stdenvs currently have this sometimes inconvenient property of respecting the platform to an extent, see this hack in firefox: https://github.com/NixOS/nixpkgs/blob/879951577d17a74f69a44e9f67f3fb65c31b9c03/pkgs/applications/networking/browsers/firefox/common.nix#L151-L153

 So arguably this PR is two things at once: adding windows llvm cross and making compiler-rt / lld detecteable independently from the platform (i.e. from the wrappers) at the same time.

shlevy replied:
OK, so do we want an `isLLD`/`isBFD` etc. passthru on all the linkers, or `family = "lld"`, `family = "bfd"`, or something else?

sternenseemann replied:
I guess currently it'd be:

* `isCctools`
* `isLld`
* `isGNU`
  * `isBfd`
  * `isGold`

Not sure if a string argument would be a good idea, so far we only have such things in the platform set.

wdyt @Ericson2314?

Ericson2314 replied:
That plan looks good to me. I am also undecided about that this strings; strings are nice to match `*Platform`, but `is*` handles "none of the above" nicely allowing "anonymous' options.

sternenseemann commented on pkgs/build-support/cc-wrapper/default.nix:
```diff
@@ -17,6 +17,7 @@
 , isGNU ? false, isClang ? cc.isClang or false, gnugrep ? null
 , buildPackages ? {}
 , libcxx ? null
+, isCompilerRT ? false
```
Isn't `useLLVM` the only way to get a compiler-rt `stdenv.cc` at the moment? `stdenv.hostPlatform.useLLVM or false` should also be a feasible replacement.

This disambiguator is probably useful regardless, but I'm wondering if we should be conservative about what we tack on here and if there's possibly a better system for this?

shlevy replied:
No, you can use this with `llvmPackages.tools.clangUseLLVM` etc.

Probably there is a better system. e.g. something like `compiler-rt` is perhaps plausibly an attribute of the platform (if it's required to be consistent across libraries and executables being linked together) whereas something like the linker really isn't (at least on platforms where there are multiple interchangeable linkers). But I'd prefer to separate out these improvements to how these are handled from the functional addition in this PR, unless there's a better current alternative.

Ericson2314 replied:
@sternenseemann yeah I think this is good too. `useLLVM` is kind of a sledge hammer when we ought to support more mix and match. I would not be opposed to specifying the *default* like we do for the linker in the platform description, but non-default compilers can do different things so we can track in the wrapper too.

sternenseemann commented on pkgs/development/compilers/llvm/13/libcxx/default.nix:
```diff
@@ -41,7 +41,13 @@ stdenv.mkDerivation rec {
+      "-DLIBCXX_ENABLE_STATIC_ABI_LIBRARY=TRUE"
+      "-DLIBCXX_CXX_ABI_LIBRARY_PATH=${libcxxabi}/lib"
+    ] ++ lib.optional (!headersOnly && libcxxabi.useLLVMUnwinder)
+      "-DLIBCXXABI_USE_LLVM_UNWINDER=ON";
```
Curious about this; `libcxx/CMakeLists.txt` claims this is only about linking the test cases which we don't run.

shlevy replied:
I got link errors without this, I wouldn't be surprised if it's only needed when `LIBCXX_ENABLE_STATIC_ABI_LIBRARY` (which is narrowly tested).

Ericson2314 replied:
In which version, @sternenseemann I don't see that in LLVM 13's `libcxx/CMakeLists.txt`.

sternenseemann replied:
I was looking at LLVM 9, but the comment is still there on `main`: https://github.com/llvm/llvm-project/blob/7e3ef7dcd2b8201dce671237857f04a1d527d580/libcxx/CMakeLists.txt#L297-L298

sternenseemann commented:
> I forward-ported the changes to 14 so they won't get lost as llvm gets upgraded

You should do `llvmPackages_git` as well then.

shlevy commented:
> > I forward-ported the changes to 14 so they won't get lost as llvm gets upgraded
> 
> You should do `llvmPackages_git` as well then.

Done

shlevy commented:
Rebased on staging.

shlevy commented on pkgs/build-support/bintools-wrapper/ld-wrapper.sh:
```diff
@@ -96,6 +96,7 @@ fi
 # Specify the target emulation if nothing is passed in ("-m" overrides this
 # environment variable). Ensures we never blindly fallback on targeting the host
 # platform.
+export LDEMULATION
```
This appears to break things. I'm building staging locally and will confirm and find an alternate approach if so.

shlevy commented:
This is back on top of master, llvm upstream rejected the LDEMULATION patch.

shlevy committed:
Add msys2 libtool patch for mingw + clang + compiler-rt

libtool upstream seems quite slow to merge...

shlevy committed:
libffi: autoreconf on Windows for libtool fix

shlevy committed:
Set a default machine type when using lld targeting Windows.

Fixes autotools checks parsing --help

shlevy committed:
libffi: Fix build for Windows with lld

shlevy committed:
gmp: autoreconf on Windows for libtool fixes.

shlevy committed:
ncurses: Fix building against ucrt

shlevy commented:
Back on staging, needed to patch libtool

shlevy commented:
@GrahamcOfBorg eval-package-list

shlevy commented:
@ofborg eval-package-list

shlevy commented:
@ofborg eval-package-list

shlevy committed:
ghcHEAD: Fix Windows cross-compilation with lld.

By specifying pkgs.libffi here instead of letting callPackage handle
it, we confuse the splicing logic and put the host libffi into
NIX_LDFLAGS_FOR_TARGET. This previously hasn't been a problem, as we
also pass an explicit configure flag pointing to the target libffi,
and so the only side-effect is a senseless -rpath flag.

ld.lld (rightly) does not recognize the -rpath flag when targeting
Windows, however, so this causes build failures.

Ericson2314 commented on pkgs/development/compilers/llvm/14/libcxxabi/default.nix:
```diff
@@ -2,9 +2,17 @@
 , cxx-headers, libunwind, version
 , enableShared ? !stdenv.hostPlatform.isStatic
-}:
+}: let
```
Nit I would do
```suggestion
}:

let
```
Using more lines but touching fiewer is good to avoid conflicts down the road.

shlevy committed:
bintools: Add isXXX flags to check linker type

shlevy commented:
@Ericson2314 @sternenseemann Any further thoughts?

sternenseemann commented:
Can you revert this merge, please? The second staging-next iteration for 22.05 has not yet started and LLVM is considered release critical, so potentially breaking changes are disallowed. See also https://github.com/NixOS/nixpkgs/issues/165792

I'd also like to have another look, didn't find the time in the last week unfortunately.

shlevy commented:
@sternenseemann Ah I didn't realize `staging-next` was restricted. This should be documented in the nixpkgs manual, currently it suggests that merging `staging` into `staging-next` when the latter is already in master (as it is now) is fine to do.

shlevy commented:
@Ericson2314 We could make the libtool patching conditional. Any strong reason to do so?
